app/models/workflow.rb

```diff
@@ -0,0 +1,39 @@
+class Workflow
+
+  attr_accessor :steps
+
+  def self.build(definition)
+    workflow = Workflow.new
+    workflow.steps = definition[:steps]
+    workflow
+  end
+
+  def start(data)
+
+    steps.each_with_index { |x, i| x[:method] = (i == 0 ? :fire : :receive) }
+
+    steps.reduce(data) do |last_event, step_data|
+
+      step = step_data[:type].constantize.new
+
+      event = step.send(step_data[:method], last_event)
+
+      event.prior_event_id = last_event.id if last_event.is_a?(Event)
+      event.data = event.data || {}
```

Comment:
makes sense. prior_event and data are part of the links between steps and generic to all steps. linking one event to another is applicable to all steps. 

`fire` and `receive` are the interfaces for event handlers, and we're only passing back the `last_event` to them. what happens inside of `fire` and `receive` are arbitrary.